File: extract_registers.py
#!/usr/bin/env python3
"""
Convert Dominus CFG/LBL files into a compact Immergas register mapping
and generate per-language label modules for the ESPHome custom component.

Writes:
- `components/immergas_modbus/immergas/labels_<lang>.py` for each language
- `immergas_registers.json` (compact PDU list)

Run from the `Immergas-Modbus` directory:
    python extract_registers.py
"""
import json
from pathlib import Path
import re
import sys

ROOT = Path(__file__).resolve().parent
CFG_PATH = (ROOT.parent / 'dominus' / 'CFG-WFC01_IM_MBUS.json').resolve()
LBL_PATH = (ROOT.parent / 'dominus' / 'LBL-WFC01_IM_MBUS.json').resolve()
OUT_PATH = ROOT / 'immergas_registers.json'
LABELS_DIR = ROOT / 'components' / 'immergas_modbus' / 'immergas'


def safe_read_json(p: Path):
    try:
        return json.loads(p.read_text(encoding='utf8'))
    except Exception as e:
        print(f'Failed to read/parse {p}: {e}', file=sys.stderr)
        sys.exit(1)


def scan_object(obj, pdus: dict):
    if obj is None:
        return
    if isinstance(obj, list):
        for v in obj:
            scan_object(v, pdus)
        return
    if not isinstance(obj, dict):
        return

    if 'pdu' in obj:
        p = obj.get('pdu')
        if p is None and p != 0:
            pass
        else:
            rec = pdus.setdefault(p, {'pdu': p, 'views': [], 'commands': [], 'messages': []})
            if 'view' in obj:
                rec['views'].append(obj['view'])
            if obj.get('action') == 'write' and 'view' in obj:
                rec['commands'].append(obj['view'])
            if 'item' in obj and 'data' in obj:
                rec['commands'].append({'item': obj['item'], 'data': obj['data']})
            if obj.get('action') in ('read', 'write'):
                rec['messages'].append({'action': obj.get('action')})

    for k, v in obj.items():
        scan_object(v, pdus)


def dedupe_list(lst):
    out = []
    seen = set()
    for x in lst:
        js = json.dumps(x, sort_keys=True)
        if js not in seen:
            seen.add(js)
            out.append(x)
    return out


def generate_label_modules(lbl_json: dict):
    # detect languages from keys like 'text1-en', 'action-it', 'comment-fr'
    langs = set()
    for entry in lbl_json.get('anomalies', []):
        for k in entry.keys():
            m = re.match(r'.+-(?P<lang>[a-z]{2})$', k)
            if m:
                langs.add(m.group('lang'))
    if not langs:
        print('No language keys found in label file')
        return []

    LABELS_DIR.mkdir(parents=True, exist_ok=True)
    generated = []
    for lang in sorted(langs):
        mapping = {}
        for entry in lbl_json.get('anomalies', []):
            code = entry.get('fault-code')
            if code is None:
                continue
            text1 = entry.get(f'text1-{lang}') or entry.get('text1-en')
            text2 = entry.get(f'text2-{lang}') or entry.get('text2-en')
            action = entry.get(f'action-{lang}') or entry.get('action-en')
            comment = entry.get(f'comment-{lang}') or entry.get('comment-en')
            mapping[code] = {
                'text1': text1,
                'text2': text2,
                'action': action,
                'comment': comment,
            }

        # write python module
        mod_path = LABELS_DIR / f'labels_{lang}.py'
        with mod_path.open('w', encoding='utf8') as f:
            f.write('"""Generated Immergas labels (language: %s)"""\n' % lang)
            f.write('\n')
            f.write('immergas_labels = {\n')
            for code, d in mapping.items():
                # ensure code is a valid literal (int or string)
                key = repr(code)
                text1 = repr(d['text1'])
                text2 = repr(d['text2'])
                action = repr(d['action'])
                comment = repr(d['comment'])
                f.write(f'    {key}: {{"text1": {text1}, "text2": {text2}, "action": {action}, "comment": {comment}}},\n')
            f.write('}\n')
        generated.append(str(mod_path))
    return generated


def main():
    cfg = safe_read_json(CFG_PATH)
    lbl = safe_read_json(LBL_PATH)

    pdus = {}
    scan_object(cfg, pdus)

    out = []
    for k, v in pdus.items():
        entry = {'pdu': k, 'views': [], 'commands': [], 'messages': []}
        for va in v.get('views', []):
            if isinstance(va, list):
                entry['views'].extend(va)
            else:
                entry['views'].append(va)
        for ca in v.get('commands', []):
            if isinstance(ca, list):
                entry['commands'].extend(ca)
            else:
                entry['commands'].append(ca)
        for m in v.get('messages', []):
            entry['messages'].append(m)
        entry['views'] = dedupe_list(entry['views'])
        entry['commands'] = dedupe_list(entry['commands'])
        entry['messages'] = dedupe_list(entry['messages'])
        out.append(entry)

    result = {
        'generated_at': __import__('datetime').datetime.utcnow().isoformat() + 'Z',
        'source_cfg': str(CFG_PATH),
        'source_lbl': str(LBL_PATH),
        'pdus': out,
        'lbl_summary': {
            'anomalies_count': len(lbl.get('anomalies', []))
        }
    }

    with OUT_PATH.open('w', encoding='utf8') as f:
        json.dump(result, f, indent=2, ensure_ascii=False)
    print('Wrote', OUT_PATH, '- PDUs found:', len(out))

    generated = generate_label_modules(lbl)
    if generated:
        print('Generated label modules:')
        for p in generated:
            print(' -', p)


if __name__ == '__main__':
    main()

File: extract_registers.js
const fs = require('fs');
const path = require('path');

const cfgPath = path.resolve(__dirname, '..', 'dominus', 'CFG-WFC01_IM_MBUS.json');
const lblPath = path.resolve(__dirname, '..', 'dominus', 'LBL-WFC01_IM_MBUS.json');

function safeReadJson(p) {
  try {
    return JSON.parse(fs.readFileSync(p, 'utf8'));
  } catch (e) {
    console.error('Failed to read/parse', p, e.message);
    process.exit(1);
  }
}

const cfg = safeReadJson(cfgPath);
const lbl = safeReadJson(lblPath);

const pdus = new Map();

function addPdu(pdu, entry) {
  if (!pdu && pdu !== 0) return;
  if (!pdus.has(pdu)) pdus.set(pdu, { pdu: pdu, views: [], commands: [], messages: [] });
  const rec = pdus.get(pdu);
  if (entry.view) rec.views.push(entry.view);
  if (entry.command) rec.commands.push(entry.command);
  if (entry.message) rec.messages.push(entry.message);
}

function scanObject(obj) {
  if (!obj || typeof obj !== 'object') return;
  if (Array.isArray(obj)) return obj.forEach(scanObject);
  // Recognize common structures
  if ('pdu' in obj) {
    const p = obj.pdu;
    if (obj.view) addPdu(p, { view: obj.view });
    if (obj.action && obj.action === 'write' && obj.view) addPdu(p, { command: obj.view });
    if (obj.item && obj.data) addPdu(p, { command: { item: obj.item, data: obj.data } });
    if (obj.action && (obj.action === 'read' || obj.action === 'write')) addPdu(p, { message: { action: obj.action } });
  }
  // Recurse
  Object.keys(obj).forEach(k => scanObject(obj[k]));
}

scanObject(cfg);

// Basic flattening + derive return types from view entries
const out = [];
for (const [k, v] of pdus.entries()) {
  const entry = { pdu: k, views: [], commands: [], messages: [] };
  v.views.forEach(viewA => {
    if (Array.isArray(viewA)) viewA.forEach(vv => entry.views.push(vv));
    else entry.views.push(viewA);
  });
  v.commands.forEach(ca => {
    if (Array.isArray(ca)) ca.forEach(cc => entry.commands.push(cc));
    else entry.commands.push(ca);
  });
  v.messages.forEach(m => entry.messages.push(m));
  // dedupe arrays
  entry.views = entry.views.filter((x, i, a) => i === a.findIndex(y => JSON.stringify(y) === JSON.stringify(x)));
  entry.commands = entry.commands.filter((x, i, a) => i === a.findIndex(y => JSON.stringify(y) === JSON.stringify(x)));
  entry.messages = entry.messages.filter((x, i, a) => i === a.findIndex(y => JSON.stringify(y) === JSON.stringify(x)));
  out.push(entry);
}

// Also include labels (a subset) to help mapping later
const result = {
  generated_at: (new Date()).toISOString(),
  source_cfg: cfgPath,
  source_lbl: lblPath,
  pdus: out,
  lbl_summary: {
    anomalies_count: Array.isArray(lbl.anomalies) ? lbl.anomalies.length : 0
  }
};

const outPath = path.resolve(__dirname, 'immergas_registers.json');
fs.writeFileSync(outPath, JSON.stringify(result, null, 2), 'utf8');
console.log('Wrote', outPath, ' - PDUs found:', out.pdu ? out.pdu.length : out.length);
